File: historical/milinda/test/org/openexi/sax/CookieTest.java
package org.openexi.sax;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.StringReader;
import java.util.ArrayList;

import org.openexi.proc.EXIDecoder;
import org.openexi.proc.common.AlignmentType;
import org.openexi.proc.common.EXIEvent;
import org.openexi.proc.common.GrammarOptions;
import org.openexi.proc.grammars.GrammarCache;
import org.openexi.proc.io.Scanner;
import org.openexi.sax.Transmogrifier;
import org.openexi.schema.EXISchema;
import org.openexi.scomp.EXISchemaFactoryErrorMonitor;
import org.openexi.scomp.EXISchemaFactoryTestUtil;
import org.xml.sax.InputSource;

import junit.framework.Assert;
import junit.framework.TestCase;

public class CookieTest extends TestCase {

  public CookieTest(String name) {
    super(name);
  }

  @Override
  protected void setUp() throws Exception {
    super.setUp();
    m_compilerErrors = new EXISchemaFactoryErrorMonitor();
  }

  @Override
  protected void tearDown() throws Exception {
    super.tearDown();
    m_compilerErrors.clear();
  }

  private EXISchemaFactoryErrorMonitor m_compilerErrors;

  private static final AlignmentType[] Alignments = new AlignmentType[] { 
    AlignmentType.bitPacked,
    AlignmentType.byteAligned,
    AlignmentType.preCompress, 
    AlignmentType.compress };
  
  ///////////////////////////////////////////////////////////////////////////
  // Test cases
  ///////////////////////////////////////////////////////////////////////////
  
  public void testCookie_01() throws Exception {
    EXISchema corpus = EXISchemaFactoryTestUtil.getEXISchema(
        (String)null, getClass(), m_compilerErrors);
    
    Assert.assertEquals(0, m_compilerErrors.getTotalCount());

    GrammarCache grammarCache = new GrammarCache(corpus, GrammarOptions.STRICT_OPTIONS);
    
    final String xmlString = "<A/>";
    
    for (AlignmentType alignment : Alignments) {
      Transmogrifier encoder = new Transmogrifier();
      EXIDecoder decoder = new EXIDecoder();
      
      encoder.setAlignmentType(alignment);
      decoder.setAlignmentType(alignment);
      
      encoder.setEXISchema(grammarCache);
      ByteArrayOutputStream baos;

      byte[] noCookie, withCookie;

      baos = new ByteArrayOutputStream();
      encoder.setOutputStream(baos);

      encoder.encode(new InputSource(new StringReader(xmlString)));
      noCookie = baos.toByteArray();

      baos = new ByteArrayOutputStream();
      encoder.setOutputStream(baos);

      encoder.setOutputCookie(true);
      encoder.encode(new InputSource(new StringReader(xmlString)));
      withCookie = baos.toByteArray();
      
      Assert.assertEquals(noCookie.length, withCookie.length - 4);

      int n_events;
      ArrayList<EXIEvent> exiEventList;

      decoder.setEXISchema(grammarCache);

      for (byte[] bts : new byte[][] { noCookie, withCookie } ) {
        decoder.setInputStream(new ByteArrayInputStream(bts));
        Scanner scanner = decoder.processHeader();
      
        exiEventList = new ArrayList<EXIEvent>();
        
        EXIEvent exiEvent;
        n_events = 0;
        while ((exiEvent = scanner.nextEvent()) != null) {
          ++n_events;
          exiEventList.add(exiEvent);
        }
        
        Assert.assertEquals(4, n_events);
        
        int pos = 0;

        exiEvent = exiEventList.get(pos++);
        Assert.assertEquals(EXIEvent.EVENT_SD, exiEvent.getEventVariety());

        exiEvent = exiEventList.get(pos++);
        Assert.assertEquals(EXIEvent.EVENT_SE, exiEvent.getEventVariety());
        Assert.assertEquals("A", exiEvent.getName());
        Assert.assertEquals("", exiEvent.getURI());

        exiEvent = exiEventList.get(pos++);
        Assert.assertEquals(EXIEvent.EVENT_EE, exiEvent.getEventVariety());
        
        exiEvent = exiEventList.get(pos++);
        Assert.assertEquals(EXIEvent.EVENT_ED, exiEvent.getEventVariety());
      }
    }
  }
  
}

File: nagasenacs/test/Nagasena/Sax/CookieTest.cs
using System;
using System.IO;
using System.Collections.Generic;
using NUnit.Framework;

using Org.System.Xml.Sax;

using EXIDecoder = Nagasena.Proc.EXIDecoder;
using AlignmentType = Nagasena.Proc.Common.AlignmentType;
using EventDescription = Nagasena.Proc.Common.EventDescription;
using EventDescription_Fields = Nagasena.Proc.Common.EventDescription_Fields;
using GrammarOptions = Nagasena.Proc.Common.GrammarOptions;
using GrammarCache = Nagasena.Proc.Grammars.GrammarCache;
using Scanner = Nagasena.Proc.IO.Scanner;
using EXISchema = Nagasena.Schema.EXISchema;
using TestBase = Nagasena.Schema.TestBase;
using EXISchemaFactoryTestUtil = Nagasena.Scomp.EXISchemaFactoryTestUtil;

namespace Nagasena.Sax {

  [TestFixture]
  public class CookieTest : TestBase {

    private static readonly AlignmentType[] Alignments = new AlignmentType[] { 
      AlignmentType.bitPacked, 
      AlignmentType.byteAligned, 
      AlignmentType.preCompress, 
      AlignmentType.compress 
    };

    ///////////////////////////////////////////////////////////////////////////
    // Test cases
    ///////////////////////////////////////////////////////////////////////////

    [Test]
    public virtual void testCookie_01() {
      EXISchema corpus = EXISchemaFactoryTestUtil.getEXISchema((string)null, this);

      GrammarCache grammarCache = new GrammarCache(corpus, GrammarOptions.STRICT_OPTIONS);

      const string xmlString = "<A/>";

      foreach (AlignmentType alignment in Alignments) {
        Transmogrifier encoder = new Transmogrifier();
        EXIDecoder decoder = new EXIDecoder();

        encoder.AlignmentType = alignment;
        decoder.AlignmentType = alignment;

        encoder.GrammarCache = grammarCache;
        MemoryStream baos;

        byte[] noCookie, withCookie;

        baos = new MemoryStream();
        encoder.OutputStream = baos;

        encoder.encode(new InputSource<Stream>(string2Stream(xmlString)));
        noCookie = baos.ToArray();

        baos = new MemoryStream();
        encoder.OutputStream = baos;

        encoder.OutputCookie = true;
        encoder.encode(new InputSource<Stream>(string2Stream(xmlString)));
        withCookie = baos.ToArray();

        Assert.AreEqual(noCookie.Length, withCookie.Length - 4);

        int n_events;
        List<EventDescription> exiEventList;

        decoder.GrammarCache = grammarCache;

        foreach (byte[] bts in new byte[][] { noCookie, withCookie }) {
          decoder.InputStream = new MemoryStream(bts);
          Scanner scanner = decoder.processHeader();

          exiEventList = new List<EventDescription>();

          EventDescription exiEvent;
          n_events = 0;
          while ((exiEvent = scanner.nextEvent()) != null) {
            ++n_events;
            exiEventList.Add(exiEvent);
          }

          Assert.AreEqual(4, n_events);

          int pos = 0;

          exiEvent = exiEventList[pos++];
          Assert.AreEqual(EventDescription_Fields.EVENT_SD, exiEvent.EventKind);

          exiEvent = exiEventList[pos++];
          Assert.AreEqual(EventDescription_Fields.EVENT_SE, exiEvent.EventKind);
          Assert.AreEqual("A", exiEvent.Name);
          Assert.AreEqual("", exiEvent.URI);

          exiEvent = exiEventList[pos++];
          Assert.AreEqual(EventDescription_Fields.EVENT_EE, exiEvent.EventKind);

          exiEvent = exiEventList[pos++];
          Assert.AreEqual(EventDescription_Fields.EVENT_ED, exiEvent.EventKind);
        }
      }
    }

  }

}